### ISSUE TITLE
[ENH] Design Discussion around pipelines and transformers which change element size. 

### ISSUE THREAD

**Issue by miraep8** [alan-turing-institute/sktime#2880]
Related to issue #2807, and potentially also #2806 and #2879. 

I think we might be able to solve some of these problems by not treating the pipeline object as a forecaster until the very last step.

More broadly, I think this might be an argument in favor of making the pipelines themselves more modular.  

Just opening this issues to kick off the discussion and connect related issues.

**Comment by fkiraly** [comment]
Could you perhaps be a bit more explicit what you mean by "not treating as forecaster until the very last step"?

This "relaxing of typing" PR could be related - is it?
https://github.com/alan-turing-institute/sktime/pull/2842

**Comment by miraep8** [comment]
Yes I would agree the "relaxing of typing" PR is relevant, thank you for linking it. 

And yes - would be happy to be more explicit!  I am still in the process of tracing all the code relevant to this, so this is probably a less than perfect understanding, but from what I could tell: 
- in the example which came up in issue 2807 `ForecastingGridSearch` we end up pre-trimming all `X` in a `fh` dependent manner via `_split` before passing it to `TransformedTargetForecaster`
- but in reality, if we are working with something like `TransformedTargetForecaster` I think it doesn't really makes sense to trim the `X` until right before the forecasting step? (Usually it isn't an issue, but it will come up with transformers that change the size of `X` like we saw) 
- so the thought was, is there a way we can delay some of these pre-processing steps that are forecaster-specific until the step in `TransformedTargetForecaster` that is right before the actual forecasting part? 

I am not actually sure that the second bullet point above actually holds.  Is this a valid point, or are their actual cases where we would need to trim the `X` in this manner for transformers as well? I think my understanding on this might be less than perfect!

**Comment by fkiraly** [comment]
> so the thought was, is there a way we can delay some of these pre-processing steps that are forecaster-specific until the step in TransformedTargetForecaster that is right before the actual forecasting part?

Isn't that what #2842 was trying to achieve, by setting pipeline tags to the "relaxed" one where no trimming takes place, instead of copying that one from the forecaster?

**Comment by fkiraly** [comment]
> but in reality, if we are working with something like `TransformedTargetForecaster` I think it doesn't really makes sense to trim the `X` until right before the forecasting step

I think your understanding is correct, but also I thought `TransformedTargetForecaster` does not trim `X`?

**Comment by fkiraly** [comment]
> I think I agree - however I was referring to the trimming that goes on in `ForecastingGridSearchCV`

I see - good catch!

So, I assume, we should not pre-trim there. Could anything bad happen if we don't, e.g., leakage?
If not, we should remove the trimming. What do you think?

Where are we currently trimming?

**Comment by miraep8** [comment]
I will make a PR attempting to fix this!  Lets see if that solves it or not!

**Comment by fkiraly** [comment]
@miraep8, quick question: does this PR exist already? Totally fine if left for later, just checking.

**Comment by miraep8** [comment]
Does not exist yet no

**Comment by miraep8** [comment]
PR has been added.  I think this PR generally seems to solve the problem, though there could be remaining issues with trimming issues in TransformedTargetForecaster.   Ideally instead of just returning the whole range of preceding data (as my solution does) we would be able to just add a buffer which depends on the amount trimmed by the pipeline, but as that solution is slightly more involved my solution just takes the naïve approach of returning the entire window.

**Comment by fkiraly** [comment]
Hm, interesting thought.

I have a generic "memory" PR for transformers here:
https://github.com/alan-turing-institute/sktime/pull/3307
which is a bit related.

How would the generic buffer work? The problem I see is that the buffer could depend on the parameters of the estimator. Simple example: `Lag`, where the buffer size would be the maximum lag.

**Comment by miraep8** [comment]
I guess a way one could do it is - all transformers have a new attribute called `buffer`.  The default for all transformers is for this required buffer to be zero, however the private method which determines the buffer size could be overwritten by the children classes and can be set to be some arbitrary function of the parameters.  

The (linear) pipeline's `buffer` would be the sum on constituent transformer's buffers.

**Comment by fkiraly** [comment]
hm, yes, that could work. Although feels a bit overengineered. What would be example cases where we would override the buffer? Could there not be just a default buffer?
